agent/engine/execcmdagent/manager_linux.go

```diff
@@ -0,0 +1,209 @@
+// The status of the ExecCommandAgent in the container is retrieved using a docker exec inspect call, using the dockerExecID
+// stored in container.GetExecCommandAgentMetadata().DockerExecID.
+//
+// If container.GetExecCommandAgentMetadata() is nil, or the ExecCommandAgent is still running, no action is taken.
+// To actually restart the ExecCommandAgent, this function invokes this instance's Start method.
+func (m *manager) RestartIfStopped(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) (bool, error) {
+	if !task.IsExecCommandAgentEnabled() {
+		return false, nil
+	}
+	execMD := container.GetExecCommandAgentMetadata()
+	if execMD == nil { // This means m.Start has never been invoked
+		return false, nil
+	}
+	resp, err := client.InspectContainerExec(ctx, execMD.DockerExecID, dockerclient.ContainerExecInspectTimeout)
+	if err != nil {
+		if _, ok := err.(*dockerapi.DockerTimeoutError); ok {
+			return false, err
+		}
+		// TODO: [ecs-exec] consider retry/adding an extra docker exec inspect API call to get more info
+		return false, err
+	}
+	if resp.Running { // Exec cmd agent is running, nothing to do
+		return false, nil
+	}
+	// Restart if not running
+	//TODO: [ecs-exec] retry only for certain exit codes?
+	seelog.Warnf("Task engine [%s]: ExecCommandAgent Process stopped (exitCode:=%d) for container %s, restarting...", task.Arn, resp.ExitCode, containerId)
+	err = m.Start(ctx, client, task, container, containerId)
+	// err == nil means start succeeded
+	return err == nil, err
+}
+
+// Start idempotently starts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// In order for the ExecCommandAgent to be successfully started, the following conditions need to be met:
+//	1) There shall be at most one ExecCommandAgent process running in the container (with the name defined in apitask.ExecCommandAgentBinName)
+//	2) If an ExecCommandAgent process is already running in the container, its PID shall be equal to container.GetExecCommandAgentMetadata().PID
+// If any of the previous conditions is violated, ExecCommandAgent's can't be verified and an error is returned.
+// If no error is returned, it can be assumed the ExecCommandAgent is started.
+func (m *manager) Start(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) error {
+	if !task.IsExecCommandAgentEnabled() {
+		return nil
+	}
+	backoff := retry.NewExponentialBackoff(m.retryMinDelay, m.retryMaxDelay, retryJitterMultiplier, retryDelayMultiplier)
+	ctx, cancel := context.WithTimeout(ctx, m.retryTimeout)
+	defer cancel()
+	var (
+		startErr error
+		attempt  int
+	)
+	retry.RetryWithBackoffCtx(ctx, backoff, func() error {
+		startErr = doStart(ctx, client, task, container, containerId)
+		if attempt == maxRetries {
+			return nil
```

Comment:
shouldn't this return an error since we reached max retries to start agent?

Reply:
This is a weird aspect of`retry.RetryWithBackoffCtx`. if max retries is reached AND there is an error, then `startErr` will store it. The return is only to break the internal loop of `retry.RetryWithBackoffCtx`. I thought about refactoring it, but it's being used in too many places. 
Other than the weird readability, it works as you can see in the UTs.

Reply:
Will see if I can use another construct of the `retry` package, but it wasn't much different.

Reply:
Changed to use `retry.RetryNWithBackoffCtx`